ubireader/ubifs/decrypt.py

```diff
@@ -0,0 +1,98 @@
+from ubireader.ubifs.defines import UBIFS_XATTR_NAME_ENCRYPTION_CONTEXT
+from ubireader.debug import error
+from cryptography.hazmat.primitives.ciphers import (
+   Cipher, algorithms, modes
+)
+
+AES_BLOCK_SIZE = algorithms.AES.block_size // 8
+
+def lookup_inode_nonce(inodes: dict, inode: dict) -> bytes:
+    # get the extended attribute 'xent' of the inode
+    if 'xent' not in inode or not inode['xent']:
+        raise ValueError(f"No xent found for inode {inode}")
+
+    for xattr_inode in inode['xent']:
+        if (xattr_inode.name == UBIFS_XATTR_NAME_ENCRYPTION_CONTEXT):
+            nonce_ino = inodes[xattr_inode.inum]['ino']
+            nonce = nonce_ino.data[-16:]
+            if len(nonce) != 16:
+                raise ValueError(f"Invalid nonce length for inode {inode}")
+            return nonce
+    
+
+def derive_key_from_nonce(master_key: bytes, nonce: bytes) -> bytes: 
+    encryptor = Cipher(
+        algorithms.AES(nonce),
+        modes.ECB(),
+    ).encryptor() 
+    derived_key = encryptor.update(master_key) + encryptor.finalize()
+    return derived_key
+
+
+def filename_decrypt(key: bytes, ciphertext: bytes):
+    
+    # using AES CTS-CBC mode not supported by pyca cryptography 
+    if len(ciphertext) > AES_BLOCK_SIZE:
+        # Cipher Text Stealing Step
+        decryptor = Cipher(
+            algorithms.AES(key[:32]),
+            modes.ECB(),
+        ).decryptor() 
+        
+        pad = (AES_BLOCK_SIZE - len(ciphertext) % AES_BLOCK_SIZE) % AES_BLOCK_SIZE
```

Comment:
```py
pad = AES_BLOCK_SIZE - len(ciphertext) % AES_BLOCK_SIZE
```

----

I would also delay `decryptor` creation, when we know, that it is needed - by moving it into the `if` block.

Reply:
We are kind of delaying it or better only create it when we need. We call the function only with
`    if ubifs.master_key is None:
        return`
surrounded.